zipline/assets/asset_writer.py

```diff
@@ -151,19 +176,70 @@ def _generate_output_dataframe(data_subset, defaults):
 
     # Get those columns which we need but
     # for which no data has been supplied.
-    need = desired_cols - cols
-
-    # Combine the users supplied data with our required columns.
-    output = pd.concat(
-        (data_subset, pd.DataFrame(
-            {k: defaults[k] for k in need},
-            data_subset.index,
-        )),
-        axis=1,
-        copy=False
+    for col in desired_cols - cols:
+        # write the default value for any missing columns
+        data_subset[col] = defaults[col]
+
+    return data_subset
+
+
+def _check_asset_group(group):
+    for colname in set(group.columns) - mapping_columns:
+        col = group[colname]
+        if len(col.unique()) != 1:
+            raise ValueError(
+                'All values must be the same for the %s column' % colname,
+            )
```

Comment:
If the error message is not enough, maybe we can update the error message to provide the context. What information do you think would help here?